spec/abstractops.html

```diff
@@ -1813,28 +1813,31 @@ <h1>
       1. Let _result_ be OrdinaryObjectCreate(*null*).
       1. Let _any_ be *false*.
       1. Let _sortedFieldNames_ be SortStringListByCodeUnit(_fieldNames_).
+      1. Let _previousProperty_ be *undefined*.
       1. For each property name _property_ of _sortedFieldNames_, do
-        1. Let _value_ be ? Get(_fields_, _property_).
-        1. If _value_ is not *undefined*, then
-          1. Set _any_ to *true*.
-          1. If _property_ is in the Property column of <emu-xref href="#table-temporal-field-requirements"></emu-xref> and there is a Conversion value in the same row, then
-            1. Let _Conversion_ be the Conversion value of the same row.
-            1. If _Conversion_ is ~ToIntegerWithTruncation~, then
-              1. Set _value_ to ? ToIntegerWithTruncation(_value_).
-              1. Set _value_ to 𝔽(_value_).
-            1. Else if _Conversion_ is ~ToPositiveIntegerWithTruncation~, then
-              1. Set _value_ to ? ToPositiveIntegerWithTruncation(_value_).
-              1. Set _value_ to 𝔽(_value_).
-            1. Else,
-              1. Assert: _Conversion_ is ~ToString~.
-              1. Set _value_ to ? ToString(_value_).
-          1. Perform ! CreateDataPropertyOrThrow(_result_, _property_, _value_).
-        1. Else if _requiredFields_ is a List, then
-          1. If _requiredFields_ contains _property_, then
-            1. Throw a *TypeError* exception.
-          1. If _property_ is in the Property column of <emu-xref href="#table-temporal-field-requirements"></emu-xref>, then
-            1. Set _value_ to the corresponding Default value of the same row.
-          1. Perform ! CreateDataPropertyOrThrow(_result_, _property_, _value_).
+        1. If _property_ is not equal to _previousProperty_, then
```

Comment:
it seems like it should either be the responsibility of the caller to ensure the list is unique (meaning, this should be an assertion) or, this AO should throw an error if the list is not unique - rather than silently ignoring it.

Reply:
I think the [argument](https://github.com/tc39/proposal-temporal/issues/2532#issue-1638914211) from the original issue is that this solution allows for better optimizability: since the fields have just been sorted, it's cheaper to remove duplicates here, rather than having the caller remove them.

Is there a particular concern that you're trying to avoid with this suggestion?

Reply:
Providing a list that has duplicates is a bug. That bug should be loudly caught as close as possible to the user input, not silently ignored deep in a chain of AOs.

Reply:
Validating the returned list from `fields()` was for a long time an open question in https://github.com/tc39/proposal-temporal/issues/1610 until Frank decided not to pursue it. I would personally be fine with throwing on duplicates returned from `fields()`. However, that would make the deduplication observable, placing a constraint on implementations that wasn't there before, so I don't know whether that fulfills @anba's original request for optimizability. I'll put this on the agenda for next week's Temporal champions meeting.

Also note that we can still get into this situation where a list with duplicates is passed to PrepareTemporalFields even without a custom calendar's `fields()` returning any duplicates, if a custom calendar would return field names like `timeZone` or `offset`, e.g. in [ToTemporalZonedDateTime](https://tc39.es/proposal-temporal/#sec-temporal-totemporalzoneddatetime), so even if we do that it still makes sense to additionally remove duplicates here. (Alternatively, disallow `timeZone` and `offset`, or append them only if they are not already present.)

Reply:
Alternatively, if a field is provided N times, why not perform the logic N times?

Either it's an instruction that should be obeyed, or it's something that should be validated, but I think it would be a very unfortunate outcome to silently attempt to "fix" anything.

Reply:
@ljharb we discussed this in the champions meeting today. I think I can summarize it as, several of us understand where your concern is coming from but we don't think it makes sense to perform the property Gets repeatedly. We were wondering if you would have time for a separate call before this is presented at the plenary, which we could coordinate through Matrix messages?

Reply:
Sure, i'm at OSSNA this week but I'm sure we can figure something out.

Reply:
Thanks @justingrant and @ljharb for the productive offline discussion! Here's a summary of what we proposed, originally written by Justin with some extra context from me.

- We'd like to throw on duplicates returned from CalendarFields, while still preserving the fact that it's not necessarily a bug if the two invocations of `fields()` return two unique lists that have some overlap when merged.
- Every invocation of CalendarFields is always followed by PrepareTemporalFields with no observable operation in between, so it doesn't matter if we throw on duplicates at the end of CalendarFields, or after sorting at the beginning of PrepareTemporalFields. Anba has indicated that throwing after sorting in PrepareTemporalFields allows optimization by only iterating the list once.
- The problem with throwing on duplicates in PrepareTemporalFields is that although CalendarFields is always followed by PrepareTemporalFields in the current spec, there's no guarantee that CalendarFields will continue to be used correctly in the future. Replace CalendarFields with two new AOs: a simple one that returns the result of PrepareTemporalFields, and a complex one that returns a record with both the names (the result of `fields()`) and the object result of PrepareTemporalFields. This new AO would need to accept an additional List of names to be added, e.g. `*offset*` and `*timeZone*`. If adding those values creates duplicates, then throw.
- Add a new parameter to PrepareTemporalFields, _duplicateBehavior_, with choices `~throw~` and `~ignore~`, in order to accommodate the behaviour of PlainYearMonth/PlainMonthDay.toPlainDate where PrepareTemporalFields is also called on the result of MergeLists,  This way, we can still implement Anba's suggestion to remove MergeLists.
- Logistics-wise, we'll change this PR to add the new parameter to PrepareTemporalFields, and throw/ignore depending on that parameter. We'd also like to address #2576 in this PR. All together, this is the normative change that we'll present to the committee on Wednesday. The refactor to replace CalendarFields is then not observable, and can be done as an editorial change without using committee time to review it.

Reply:
Opened #2630 for the follow up editorial refactor.

Reply:
This is now done in https://github.com/tc39/proposal-temporal/pull/2823